src/Workspaces/Core/Portable/Rename/Renamer.RenameDocumentAction.cs

```diff
@@ -0,0 +1,53 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections.Immutable;
+using System.Globalization;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Options;
+
+namespace Microsoft.CodeAnalysis.Rename
+{
+    public static partial class Renamer
+    {
+        /// <summary>
+        /// Individual action from RenameDocument APIs in <see cref="Renamer"/>.
+        /// 
+        /// See <see cref="RenameDocumentActionSet" />
+        /// </summary>
+        public abstract class RenameDocumentAction
+        {
+            private readonly ImmutableArray<ErrorResource> _errorStringKeys;
+            protected OptionSet OptionSet { get; }
+
+            public virtual ImmutableArray<string> GetErrors(CultureInfo? culture = null)
+                => _errorStringKeys.SelectAsArray(s => string.Format(WorkspacesResources.ResourceManager.GetString(s.FormatString, culture ?? WorkspacesResources.Culture)!, s.Arguments));
```

Comment:
i'm surprised this is virtual.

Reply:
Theoretically if someone wanted to implement this abstract class they could. I didn't want to lock them into a GetErrors implementation that they couldn't use. Instead, they could provide their own.

Reply:
oh no.  no no no  :)

do not have speculative inheritance paths.  in fact, this type should make sure to have an internal/private constructor so that it can *not* be publicly subclassed.

We only expose actual inheritability when we've spent the time to design for that.  It's hard to do and should only be done if necessary :)

Reply:
cool. I'll remove the virtual. I wasn't 100% sure that internal constructor would mean that the type couldn't be subclassed. The constructor is internal so we should be good